snap/squashfs2/squashfs2.go

```diff
@@ -0,0 +1,531 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * https://www.kernel.org/doc/html/v5.8/filesystems/squashfs.html
+ */
+
+package squashfs2
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+const (
+	// https://github.com/plougher/squashfs-tools/blob/master/squashfs-tools/squashfs_fs.h#L289
+	superblockSize         = 96
+	metadataBlockSize      = 8192
+	directoryMaxEntryCount = 256
+
+	superBlockUncompressedInodes    = 0x1
+	superBlockUncompressedData      = 0x2
+	superBlockUncompressedFragments = 0x8
+	superBlockNoFragments           = 0x10
+	superBlockAlwaysFragments       = 0x20
+	superBlockDublicates            = 0x40
+	superBlockExportable            = 0x80
+	superBlockUncompressedXattrs    = 0x100
+	superBlockNoXattrs              = 0x200
+	superBlockCompressorOptions     = 0x400
+	superBlockUncompressedIds       = 0x800
+
+	// Inode types supported by squashfs
+	inodeTypeDirectory         = 1
+	inodeTypeFile              = 2
+	inodeTypeSymlink           = 3
+	inodeTypeBlockDev          = 4
+	inodeTypeCharDev           = 5
+	inodeTypeFifo              = 6
+	inodeTypeSocket            = 7
+	inodeTypeExtendedDirectory = 8
+	inodeTypeExtendedFile      = 9
+	inodeTypeExtendedSymlink   = 10
+	inodeTypeExtendedBlockDev  = 11
+	inodeTypeExtendedCharDev   = 12
+	inodeTypeExtendedFifo      = 13
+	inodeTypeExtendedSocket    = 14
+
+	// Compression types supported by squashfs
+	compressionZlib = 1
+	compressionLzma = 2
+	compressionLzo  = 3
+	compressionXz   = 4
+	compressionLz4  = 5
+	compressionZstd = 6
+)
+
+var (
+	// magic is the magic prefix of squashfs snap files.
+	magic = [4]byte{'h', 's', 'q', 's'}
```

Comment:
I do a comparison between two [4]byte arrays in the superblock parsing, it won't work if I don't specify the size. It will crap out over arrays being different sizes. This is the reason for specifying the size

Reply:
No strong opinion about that (IMHO it's fine for predefined byte sequence like this), but just FYI in case it's useful, you can use `bytes.Compare(foo[:4], bar[:4])` in such cases to compare slices.

Reply:
Ah thanks for informing me about the bytes.Compare as well - didn't know this!